#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-39_noepg.dpatch from version the vdr-noepgmenu plugin 0.0.6.beta3
## http://winni.vdr-developer.org/noepgmenu/downloads/beta
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: The patch allows to disable normal epg update for specified channels.
## DP: This is useful if you get epg data of the channels from external sources.

@DPATCH@
diff -urNad vdr-1.6.0~/config.c vdr-1.6.0/config.c
--- vdr-1.6.0~/config.c	2008-05-11 03:35:16.000000000 +0200
+++ vdr-1.6.0/config.c	2008-05-11 03:35:44.000000000 +0200
@@ -339,11 +339,20 @@
   ShowRecLength = 0;
   ShowProgressBar = 0;
   MenuCmdPosition = 0;
+  noEPGMode=0;
+  noEPGList=strdup("");
+}
+
+cSetup::~cSetup()
+{
+  free(noEPGList);
 }
 
 cSetup& cSetup::operator= (const cSetup &s)
 {
   memcpy(&__BeginData__, &s.__BeginData__, (char *)&s.__EndData__ - (char *)&s.__BeginData__);
+  free(noEPGList);
+  noEPGList = strdup(s.noEPGList);
   return *this;
 }
 
@@ -504,6 +513,11 @@
   else if (!strcasecmp(Name, "FontOsdSize"))         FontOsdSize        = atoi(Value);
   else if (!strcasecmp(Name, "FontSmlSize"))         FontSmlSize        = atoi(Value);
   else if (!strcasecmp(Name, "FontFixSize"))         FontFixSize        = atoi(Value);
+  else if (!strcasecmp(Name, "noEPGMode"))           noEPGMode          = atoi(Value);
+  else if (!strcasecmp(Name, "noEPGList")) {
+    free(noEPGList);
+    noEPGList=strdup(Value ? Value : "");
+  }
   else if (!strcasecmp(Name, "MaxVideoFileSize"))    MaxVideoFileSize   = atoi(Value);
   else if (!strcasecmp(Name, "SplitEditedFiles"))    SplitEditedFiles   = atoi(Value);
   else if (!strcasecmp(Name, "MinEventTimeout"))     MinEventTimeout    = atoi(Value);
@@ -599,6 +613,8 @@
   Store("FontOsd",            FontOsd);
   Store("FontSml",            FontSml);
   Store("FontFix",            FontFix);
+  Store("noEPGMode",          noEPGMode);
+  Store("noEPGList",          noEPGList);
   Store("FontOsdSize",        FontOsdSize);
   Store("FontSmlSize",        FontSmlSize);
   Store("FontFixSize",        FontFixSize);
diff -urNad vdr-1.6.0~/config.h vdr-1.6.0/config.h
--- vdr-1.6.0~/config.h	2008-05-11 03:35:16.000000000 +0200
+++ vdr-1.6.0/config.h	2008-05-11 03:35:16.000000000 +0200
@@ -280,6 +280,7 @@
   int FontOsdSize;
   int FontSmlSize;
   int FontFixSize;
+  int noEPGMode;
   int MaxVideoFileSize;
   int SplitEditedFiles;
   int MinEventTimeout, MinUserInactivity;
@@ -299,7 +300,9 @@
   int EmergencyExit;
   int ShowRecDate, ShowRecTime, ShowRecLength, ShowProgressBar, MenuCmdPosition;
   int __EndData__;
+  char *noEPGList; // pointer not to be flat-copied
   cSetup(void);
+  ~cSetup();
   cSetup& operator= (const cSetup &s);
   bool Load(const char *FileName);
   bool Save(void);
diff -urNad vdr-1.6.0~/eit.c vdr-1.6.0/eit.c
--- vdr-1.6.0~/eit.c	2008-05-11 03:35:16.000000000 +0200
+++ vdr-1.6.0/eit.c	2008-05-11 03:35:16.000000000 +0200
@@ -22,8 +22,28 @@
 class cEIT : public SI::EIT {
 public:
   cEIT(cSchedules *Schedules, int Source, u_char Tid, const u_char *Data, bool OnlyRunningStatus = false);
+
+private:
+  bool allowedEPG(tChannelID kanalID);
   };
 
+bool cEIT::allowedEPG(tChannelID kanalID) {
+  bool rc;
+
+  if (Setup.noEPGMode == 1) {
+     rc=false;
+     if (strstr(::Setup.noEPGList,kanalID.ToString())!=NULL)
+        rc=true;
+     }
+  else {
+     rc=true;
+     if (strstr(::Setup.noEPGList,kanalID.ToString())!=NULL)
+        rc=false;
+     }
+
+  return rc;
+}
+
 cEIT::cEIT(cSchedules *Schedules, int Source, u_char Tid, const u_char *Data, bool OnlyRunningStatus)
 :SI::EIT(Data, false)
 {
@@ -35,6 +55,12 @@
   if (!channel)
      return; // only collect data for known channels
 
+  // only use epg from channels not blocked by noEPG-patch
+  tChannelID kanalID;
+  kanalID=channel->GetChannelID();
+  if (!allowedEPG(kanalID))
+    return;
+
   cSchedule *pSchedule = (cSchedule *)Schedules->GetSchedule(channel, true);
 
   bool Empty = true;
diff -urNad vdr-1.6.0~/menu.c vdr-1.6.0/menu.c
--- vdr-1.6.0~/menu.c	2008-05-11 03:35:16.000000000 +0200
+++ vdr-1.6.0/menu.c	2008-05-11 03:35:16.000000000 +0200
@@ -2604,6 +2604,7 @@
 
 class cMenuSetupEPG : public cMenuSetupBase {
 private:
+  const char *noEPGModes[2];
   int originalNumLanguages;
   int numLanguages;
   void Setup(void);
@@ -2626,6 +2627,9 @@
 {
   int current = Current();
 
+  noEPGModes[0]=tr("Blacklist");
+  noEPGModes[1]=tr("Whitelist");
+
   Clear();
 
   Add(new cMenuEditIntItem( tr("Setup.EPG$EPG scan timeout (h)"),      &data.EPGScanTimeout));
@@ -2644,6 +2648,7 @@
   Add(new cMenuEditBoolItem(tr("Setup.EPG$extern double Epg entry"),   &data.DoubleEpgAction, "adjust", "delete"));
   Add(new cMenuEditBoolItem(tr("Setup.EPG$Mix intern and extern EPG"), &data.MixEpgAction));
   Add(new cMenuEditBoolItem(tr("Setup.EPG$Disable running VPS event"), &data.DisableVPS));
+  Add(new cMenuEditStraItem(tr("Setup.EPG$Mode noEPG-Patch"),          &data.noEPGMode, 2, noEPGModes));
 
   SetCurrent(Get(current));
   Display();
